internal/app/machined/pkg/controllers/k8s/node_label_spec.go

```diff
@@ -0,0 +1,326 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	"github.com/talos-systems/talos/pkg/kubernetes"
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1/machine"
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+	"github.com/talos-systems/talos/pkg/machinery/resources/k8s"
+	"github.com/talos-systems/talos/pkg/machinery/resources/secrets"
+)
+
+// NodeLabelsK8sClientIface is the minimal interface needed fo manage the node labels.
+type NodeLabelsK8sClientIface interface {
+	Close() error
+	GetNodeLabels(ctx context.Context, nodeName string) (map[string]string, map[string]struct{}, error)
+	UpdateNodeLabels(ctx context.Context, nodeName string, labelsToSet map[string]string, labelsToRemove map[string]struct{}) (err error)
+}
+
+// NodeLabelSpecController creates k8s.NodeLabelSpec's and applies them to the k8s Node object.
+type NodeLabelSpecController struct {
+	GetK8sClientFunc func(ctx context.Context, r controller.Runtime) (NodeLabelsK8sClientIface, error)
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *NodeLabelSpecController) Name() string {
+	return "k8s.NodeLabelSpecController"
+}
+
+// KubernetesClientDependencies holds the dependencies needed for a controller to create a kubernetes client on control plane or worker nodes.
+var KubernetesClientDependencies = []controller.Input{
```

Comment:
I'm still lost on the complexity here... why do we have so many controllers changing things?

let's pick a simple version which will do its job (two controllers):

1. convert machine config to a list of node labels to have (this should be NodeLabelConfig, or NodeLabelSpec to be consistent with naming).
2. controller which watches NodeLabelConfig and on change builds Kubernetes client, gets `Node`, and compares config (spec) with the actual node state. If label is being added, add it to the special annotation which says that it's managed by Talos; if label is being updated, and it's already in the annotation, update it, otherwise error out; if the label is being removed, and it's already in the annotation, remove it and remove from annotation, otherwise error out.

Done. This is pretty good solution with the only downside it doesn't handle the case if someone goes and updates the labels themselves.

I would probably stop here, but if we don't want to stop, then we need another controller:

1. Controller which watches/polls Node and publishes NodeLabelStatus.
2. Feed NodeLabelStatus to the controller which manages labels, this way (without even reading NodeLabelStatus) it will reconcile things on each node label change.